lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart

```diff
@@ -0,0 +1,71 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:ui/src/engine.dart';
+import 'package:ui/ui.dart' as ui;
+
+/// Tracks the [FlutterView]s focus changes.
+final class ViewFocusBinding {
+  ViewFocusBinding._();
+
+  static ViewFocusBinding? _instance;
+
+  /// Initializes the [ViewFocusBinding] singleton.
+  static void initialize() {
+    _instance ??= ViewFocusBinding._().._init();
+  }
+
+  final EnginePlatformDispatcher _dispatcher =
+      EnginePlatformDispatcher.instance;
+
+  void _init() {
+    domDocument.body?.addEventListener(_focusin, _focusChangeHandler, true);
+    domDocument.body?.addEventListener(_focusout, _focusChangeHandler, true);
+    registerHotRestartListener(_dispose);
+  }
+
+  void _dispose() {
+    domDocument.body?.removeEventListener(_focusin, _focusChangeHandler, true);
+    domDocument.body?.removeEventListener(_focusout, _focusChangeHandler, true);
+    _instance = null;
+  }
+
+  int? _latestFocusedViewId;
+  late final DomEventListener _focusChangeHandler = createDomEventListener((DomEvent event) {
+    final int? viewId = _findViewIdOrNull(domDocument.activeElement);
+    if (viewId == _latestFocusedViewId) {
+      return;
+    }
+
+    late final ui.ViewFocusEvent event;
+    if (viewId == null) {
+      event = ui.ViewFocusEvent(
+        viewId: _latestFocusedViewId!,
+        state: ui.ViewFocusState.unfocused,
+        direction: ui.ViewFocusDirection.undefined,
+      );
+    } else {
+      event = ui.ViewFocusEvent(
+        viewId: viewId,
+        state: ui.ViewFocusState.focused,
+        direction: ui.ViewFocusDirection.forward,
+      );
+    }
+    _latestFocusedViewId = viewId;
+    _dispatcher.invokeOnViewFocusChange(event);
+  });
+}
+
+const String _focusin = 'focusin';
+const String _focusout = 'focusout';
```

Comment:
Move these inside the ViewFocusBinding class, and make them `static`, so they're not lost in the middle of the library